reana_client/utils.py

```diff
@@ -123,7 +124,6 @@ def _prepare_kwargs(reana_yaml):
     try:
         with open(filepath) as f:
             reana_yaml = yaml.load(f.read(), Loader=yaml.FullLoader)
-
         workflow_type = reana_yaml["workflow"]["type"]
```

Comment:
E.g. If we have `typ` instead of `type` we'll have an exception here without even reaching the JSON Schema validation (`validate(reana_yaml, reana_yaml_schema)`). Perhaps we can move the validation to be performed before and detect schema problems earlier?

```diff
diff --git a/reana.yaml b/reana.yaml
index 9fcfa12..4cf462c 100644
--- a/reana.yaml
+++ b/reana.yaml
@@ -11,7 +11,7 @@ inputs:
     year_min: 1500
     year_max: 2012
 workflow:
-  type: serial
+  typ: serial
   specification:
     steps:
       - environment: 'reanahub/reana-env-jupyter:2.0.0'
```

```console
$ reana-client validate    
==> ERROR: Something went wrong when trying to validate ~/code/reanahub/reana-demo-worldpopulation/reana.yaml
```
I have no hints on what is wrong.

Reply:
Following the IRL conversation, I meant something like this:

```diff
diff --git a/reana_client/utils.py b/reana_client/utils.py
index 5bc4312..e840648 100644
--- a/reana_client/utils.py
+++ b/reana_client/utils.py
@@ -124,7 +124,16 @@ def load_reana_spec(
     try:
         with open(filepath) as f:
             reana_yaml = yaml.load(f.read(), Loader=yaml.FullLoader)
-        workflow_type = reana_yaml["workflow"]["type"]
+        workflow_type = reana_yaml["workflow"].get("type")
+        if not skip_validation:
+            display_message(
+                "Verifying REANA specification file... {filepath}".format(
+                    filepath=filepath
+                ),
+                msg_type="info",
+            )
+            _validate_reana_yaml(reana_yaml)
+            validate_parameters(workflow_type, reana_yaml)
         reana_yaml["workflow"]["specification"] = load_workflow_spec(
             workflow_type,
             reana_yaml["workflow"].get("file"),
@@ -139,16 +148,6 @@ def load_reana_spec(
                     f, Loader=yaml.FullLoader
                 )
 
-        if not skip_validation:
-            display_message(
-                "Verifying REANA specification file... {filepath}".format(
-                    filepath=filepath
-                ),
-                msg_type="info",
-            )
-            _validate_reana_yaml(reana_yaml)
-            validate_parameters(workflow_type, reana_yaml)
-
         if not skip_validate_environments:
             display_message(
                 "Verifying environments in REANA specification file...",
@@ -203,6 +202,9 @@ def _validate_reana_yaml(reana_yaml):
         )
         raise e
     except ValidationError as e:
+        import wdb
+
+        wdb.set_trace()
         logging.info("Invalid REANA specification: {error}".format(error=e.message))
         if e.validator == "additionalProperties":
             invalid_section = (
```

So then the output is more informative:
```console
$ reana-client validate 
==> Verifying REANA specification file... /Users/marco/code/reanahub/reana-demo-worldpopulation/reana.yaml
==> ERROR: /Users/marco/code/reanahub/reana-demo-worldpopulation/reana.yaml is not a valid REANA specification:
==> ERROR:  The `reana.yaml` directive "typ" is not valid.
```

Reply:
I agree to move the `_validate_reana_yaml` but `validate_parameters` would need to load the specification file. I would rather separate them